crates/pera-core/src/dwallet_mpc/mpc_session.rs

```diff
@@ -143,12 +132,132 @@ impl DWalletMPCSession {
+                let public_input = bcs::from_bytes(&self.public_input)?;
+                crate::dwallet_mpc::advance::<PresignFirstParty>(
+                    session_id,
+                    self.party_id,
+                    &self.weighted_threshold_access_structure,
+                    self.pending_messages.clone(),
+                    public_input,
+                    (),
+                )
+            }
+            MPCRound::PresignSecond(..) => {
+                let public_input = bcs::from_bytes(&self.public_input)?;
+                crate::dwallet_mpc::advance::<PresignSecondParty>(
+                    session_id,
+                    self.party_id,
+                    &self.weighted_threshold_access_structure,
+                    self.pending_messages.clone(),
+                    public_input,
+                    (),
+                )
+            }
+            MPCRound::Sign(..) => {
+                let public_input = bcs::from_bytes(&self.public_input)?;
+                crate::dwallet_mpc::advance::<SignFirstParty>(
+                    session_id,
+                    self.party_id,
+                    &self.weighted_threshold_access_structure,
+                    self.pending_messages.clone(),
+                    public_input,
+                    self.decryption_share.clone(),
+                )
+            }
+            MPCRound::NetworkDkg(key_scheme, _) => advance_network_dkg(
+                session_id,
+                &self.weighted_threshold_access_structure,
+                self.party_id,
+                &self.public_input,
+                key_scheme,
+                self.pending_messages.clone(),
+                bcs::from_bytes(
+                    &self
+                        .private_input
+                        .clone()
+                        .ok_or(DwalletMPCError::MissingMPCPrivateInput)?,
+                )?,
+            ),
+            MPCRound::EncryptedShareVerification(verification_data) => {
+                match verify_encrypted_share(verification_data) {
+                    Ok(_) => Ok(AsynchronousRoundResult::Finalize {
+                        public_output: vec![],
+                        private_output: vec![],
+                        malicious_parties: vec![],
+                    }),
+                    Err(err) => Err(err),
+                }
+            }
+            MPCRound::EncryptionKeyVerification(verification_data) => {
+                verify_encryption_key(verification_data)
+                    .map(|_| AsynchronousRoundResult::Finalize {
+                        public_output: vec![],
+                        private_output: vec![],
+                        malicious_parties: vec![],
+                    })
+                    .map_err(|err| err)
+            }
+            MPCRound::BatchedPresign(..) | MPCRound::BatchedSign(..) => {
+                unreachable!("advance should never be called on a batched session")
```

Comment:
Remind self to check this can never be reached even if other validators ir users behave maliciously